flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java

```diff
@@ -396,11 +400,155 @@ public void testPredefinedOptions() throws Exception {
+				assertEquals(CompactionStyle.LEVEL, columnOptions.compactionStyle());
+				assertTrue(columnOptions.levelCompactionDynamicLevelBytes());
+				assertEquals(8 * SizeUnit.MB, columnOptions.targetFileSizeBase());
+				assertEquals(128 * SizeUnit.MB, columnOptions.maxBytesForLevelBase());
+				assertEquals(4, columnOptions.maxWriteBufferNumber());
+				assertEquals(2, columnOptions.minWriteBufferNumberToMerge());
+				assertEquals(64 * SizeUnit.MB, columnOptions.writeBufferSize());
+
+				BlockBasedTableConfig tableConfig = (BlockBasedTableConfig) columnOptions.tableFormatConfig();
+				assertEquals(4 * SizeUnit.KB, tableConfig.blockSize());
+				assertEquals(512 * SizeUnit.MB, tableConfig.blockCacheSize());
+			}
+		}
+	}
+
+	/**
+	 * This test verify configurations from flink-conf.yaml would be overrode by programmatically configured.
+	 */
+	@Test
+	public void testConfigOptions() throws Exception  {
+		Configuration configuration = new Configuration();
+		configuration.setString(RocksDBOptions.PREDEFINED_OPTIONS, PredefinedOptions.SPINNING_DISK_OPTIMIZED.name());
+		configuration.setString(RocksDBConfigurableOptions.MAX_WRITE_BUFFER_NUMBER, "4");
+
+		String checkpointPath = tempFolder.newFolder().toURI().toString();
+		RocksDBStateBackend rocksDbBackend = new RocksDBStateBackend(checkpointPath);
+
+		// we re-configured predefined options as 'PredefinedOptions.DEFAULT' programmatically.
+		rocksDbBackend.setPredefinedOptions(PredefinedOptions.DEFAULT);
+
+		ConfigurableOptionsFactory optionsFactory = new ConfigurableOptionsFactory().setMaxWriteBufferNumber(3);
+		// we re-configured max write-buffer number as '3' programmatically.
+		rocksDbBackend.setOptions(optionsFactory);
+
+		rocksDbBackend = rocksDbBackend.configure(configuration);
+
+		assertEquals(PredefinedOptions.DEFAULT, rocksDbBackend.getPredefinedOptions());
+
+		try (ColumnFamilyOptions columnFamilyOptions = rocksDbBackend.getColumnOptions()){
+			assertEquals(3, columnFamilyOptions.maxWriteBufferNumber());
+		}
+	}
+
 	@Test
 	public void testOptionsFactory() throws Exception {
 		String checkpointPath = tempFolder.newFolder().toURI().toString();
 		RocksDBStateBackend rocksDbBackend = new RocksDBStateBackend(checkpointPath);
 
+		// 1. verify configure user-defined options factory through flink-conf.yaml
+		Configuration config = new Configuration();
+		config.setString(RocksDBOptions.OPTIONS_FACTORY.key(), TestOptionsFactory.class.getName());
+
+		rocksDbBackend = rocksDbBackend.configure(config);
+
+		final Environment env = getMockEnvironment(tempFolder.newFolder());
+		// call createKeyedStateBackend to instantiate the user-defined options factory.
+		createKeyedStateBackend(rocksDbBackend, env);
+
+		assertNotNull(rocksDbBackend.getOptions());
+		try (ColumnFamilyOptions colCreated = rocksDbBackend.getColumnOptions()) {
+			assertEquals(CompactionStyle.UNIVERSAL, colCreated.compactionStyle());
```

Comment:
Why do we validate the successful setting of the options factory in such a complicated way? Wouln't it be enough to just check that `getOptions` returns an instance of `TestOptionsFactory`? There should already be another test to validate the order that setter overrides flink-conf and so on. This would keep the test more targetet or am I missing something?

Reply:
This is because `getOptions` actually returns the `UserConfiguredOptionsFactory` warping the `TestOptionsFactory`. I think if we could solve the `ClassLoader` problem, we could just instantiate the options factory directly and verify that `getOptions` returns an instance of `TestOptionsFactory`.

Reply:
I will try to double-check about the classloader. My feeling is that this code should run on the client and that is already in a user code context, so that we could use the context classloader of the current thread. But I prefer to first really check this and then will get back to you.